src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs

```diff
@@ -521,19 +521,17 @@ private RegexNode ScanReplacement()
                 int startpos = _pos;
 
                 _pos = _pattern.IndexOf('$', _pos);
-                if (_pos == -1)
+                if (_pos < 0)
                     _pos = _pattern.Length;
 
                 AddToConcatenate(startpos, _pos - startpos, isReplacement: true);
 
                 if (_pos < _pattern.Length)
                 {
-                    if (_pattern[_pos++] == '$')
-                    {
-                        _unit = ScanDollar();
-                    }
+                    _pos++;
+                    _unit = ScanDollar();
 
-                    _concatenation.AddChild(_unit!);
+                    _concatenation.AddChild(_unit);
```

Comment:
Can this be:
```C#
_pos++;
_concatenation.AddChild(ScanDollar());
_unit = null;
```
?

Reply:
yes, changed, but in general the to and fro with _unit and _concatenation and _group etc can probably be cleaned up in future. Eg., I think each time you add a child to a concatenation, you null _unit, but all the callers do it. etc